### PR TITLE
changes from fv3net

### PR DESCRIPTION
**Pull request by nbren12**
This PR contains a new feature developed in an the fv3net PR 
https://github.com/VulcanClimateModeling/fv3net/pull/1139#pullrequestreview-631814431

I tried to write a clean git history, so I recommend merging using
a method that preservers it e.g. rebase-and-merge or traditional merge.

Changes from our last discussion:

- I grouped the chunks and regions data into a data container class called
  `_ValidWorkPlan`.

### CODE REVIEW

File: test_xpartition.py [hunk 1]
```diff
@@ -145,3 +145,49 @@ def test_dataset_mappable_write(tmpdir, ds, ranks):
...
+    mapper = ds.z.partition.map(store, ranks=3, dims=["x"], func=func, data=ds)
+    for rank in mapper:
+        mapper.write(rank)
+
+    written = xr.open_zarr(store)
+    xr.testing.assert_identical(func(ds), written)
+
+
+def test_partition_partition():
+    # Partitions have two qualitities which we test using the a data-array that
```
Comment (spencerkclark) [inline test_xpartition.py]:
```suggestion
    # Partitions have two qualities which we test using a DataArray that
```

File: test_xpartition.py [hunk 2]
```diff
@@ -145,3 +145,49 @@ def test_dataset_mappable_write(tmpdir, ds, ranks):
...
+    # has all unique values
+    ds = xr.Dataset(
+        {"z": (["x", "y"], np.arange(50).reshape((5, 10)), {"an": "attr"})},
```
Comment (spencerkclark) [inline test_xpartition.py]: nit: I think the attribute does not add anything to this particular test (it's useful in the one above).
```suggestion
        {"z": (["x", "y"], np.arange(50).reshape((5, 10)))},
```

File: xpartition.py [hunk 3]
```diff
@@ -271,3 +301,77 @@ def mappable_write(
...
+        region = self.plan.input_partition[0]
+        iData = self.data.isel(region)
+        iOut = self.func(iData)
+        full_indexers = {dim: self.data[dim] for dim in self.dims}
+
+        dims_without_coords = (set(iOut.dims) - set(iOut.indexes)) & set(self.dims)
+        for dim in dims_without_coords:
+            iOut = iOut.assign_coords({dim: iOut[dim]})
+
+        schema = zeros_like(iOut.reindex(full_indexers), chunks=self.plan.output_chunks)
```
Comment (spencerkclark) [inline xpartition.py:355]: Is there a reason `xarray.zeros_like` does not work here?

File: setup.cfg [hunk 4]
```diff
@@ -16,13 +16,15 @@ classifiers =
 
 [options]
 packages = find:
+py_modules = xpartition
 zip_safe = False  # https://mypy.readthedocs.io/en/latest/installed_packages.html
 include_package_data = True
 python_requires = >=3.6
```
Comment (spencerkclark) [inline setup.cfg:22]: If you don't mind updating this too -- I think xarray requires at least Python 3.7 these days (plus this PR uses dataclasses):
```suggestion
python_requires = >=3.7
```

**Review by spencerkclark — APPROVED** [review]
Thanks for the nice additions!  Just had a couple minor comments / questions.

File: xpartition.py (hunk 3 above)
Comment (nbren12) [reply xpartition.py:355]: It has no "chunks" arg: http://xarray.pydata.org/en/stable/generated/xarray.zeros_like.html

Comment (spencerkclark) [reply xpartition.py:355]: Would this do something undesired?
```suggestion
        schema = xr.zeros_like(iOut.reindex(full_indexers).chunk(self.plan.output_chunks))
```

File: setup.cfg (hunk 4 above)
Comment (nbren12) [reply setup.cfg:22]: xarray 0.16.2 supports python 3.6, and we still use it in the prognostic run image. I fixed the setup.cfg so it works on py3.6 as well, and tested via tox.

File: xpartition.py (hunk 3 above)
Comment (nbren12) [reply xpartition.py:355]: I think you suggestion is good enough, although the dask graph of the schema will have an extra layer of rechunking operations.

Comment (nbren12) [reply xpartition.py:355]: I implemented your suggestion.

Comment (spencerkclark) [reply xpartition.py:355]: Thanks -- I think `iOut.reindex(full_indexers).chunk(self.plan.output_chunks)` will have an extra layer of rechunking operations, but `schema` should just be constructed from the final chunk structure if I'm not mistaken.

**Pull request merged**